homeassistant/components/wilight/parent_device.py

```diff
@@ -0,0 +1,110 @@
+"""The WiLight Device integration."""
+import asyncio
+import logging
+
+import pywilight
+import requests
+
+from homeassistant.const import CONF_HOST, EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import async_dispatcher_send
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class WiLightParent:
+    """Manages a single WiLight Parent Device."""
+
+    def __init__(self, hass, config_entry):
+        """Initialize the system."""
+        self._host = config_entry.data[CONF_HOST]
+        self._hass = hass
+        self._api = None
+
+    @property
+    def host(self):
+        """Return the host of this parent."""
+        return self._host
+
+    @property
+    def api(self):
+        """Return the api of this parent."""
+        return self._api
+
+    async def async_setup(self):
+        """Set up a WiLight Parent Device based on host parameter."""
+        host = self._host
+        hass = self._hass
+
+        api_device = await hass.async_add_executor_job(create_api_device, host, 45995)
+
+        if api_device is None:
+            return False
+
+        @callback
+        def disconnected():
+            # Schedule reconnect after connection has been lost.
+            _LOGGER.warning("WiLight %s disconnected", api_device.device_id)
+            async_dispatcher_send(
+                hass, f"wilight_device_available_{api_device.device_id}", False
+            )
+
+        @callback
+        def reconnected():
+            # Schedule reconnect after connection has been lost.
+            _LOGGER.warning("WiLight %s reconnect", api_device.device_id)
+            async_dispatcher_send(
+                hass, f"wilight_device_available_{api_device.device_id}", True
+            )
+
+        async def connect(api_device):
+            # Set up connection and hook it into HA for reconnect/shutdown.
+            _LOGGER.debug("Initiating connection to %s", api_device.device_id)
+
+            client = await api_device.config_client(
```

Comment:
This task can take take forever if the connection can't be established. We should either have a timeout or let the library handle the connection and not hold up the integration setup for this. In the latter case we need to schedule the task with `asyncio.create_task`.

Reply:
I noticed when looking at the library that it uses `time.sleep` in a reconnect function. I don't see that function being used anywhere, but please make sure that we don't use `time.sleep` in the event loop as it will block it.

Reply:
The initial connection has a timeout, but the reconnection did have a sleep.
I'm updating the API and exchanging sleep for:
    self._reconnect_timer = self._loop.call_later (self._reconnect_interval, self.do_reconnect)
The self.do_reconnect procedure has a timeout.

Reply:
Looks like the `while True:` can be removed since there's a `break` at the end:
https://github.com/leofig-rj/pywilight/blob/71fae849d5464a444bef3bf9671046a8f70332b1/pywilight/wilight_device/protocol.py#L488

Reply:
You are correct, I will remove. I was testing it and then I was going to review it.

Reply:
Done in Version 0.0.65